Summary of speedup shapes layer project (#10)
napari/island-dispatch: source/blog/triangulation_summary.md

---
blogpost: true
date: Jun 18, 2025
author: Grzegorz Bokota
category: Advertisement
language: English
---

# Summary of speedup of Shapes layer

> With a big thank you to  the SpatialData team, who decided to sponsor this work over the past last half year, Grzegorz Bokota, one of the Core-Developers of napari, has been working on speedup of Shapes layer. The work has been done in the context of the napari project, and the results are already available in the latest release of napari.

Reviewer: ```suggestion
With a big thank you to  the SpatialData team, who decided to sponsor this work over the past last half year, Grzegorz Bokota, one of the Core-Developers of napari, has been working on speedup of the Shapes layer. The work has been done in the context of the napari project, and the results are already available in the latest release of napari.
```

* [#10](https://github.com/napari/bermuda/pull/10) - Implement monotone polygon triangulation in Rust
* [#14](https://github.com/napari/bermuda/pull/14) - Deduplicate of edges for edge and face triangulation
* [#17](https://github.com/napari/bermuda/pull/17) - Sweeping line algorithm for triangulation in Rust, Python API for face and edge triangulation in one call
* [#27](https://github.com/napari/bermuda/pull/27) - Python API for face triangulation without edges
* [#30](https://github.com/napari/bermuda/pull/30) - Checking if polygon is convex 
* [#32](https://github.com/napari/bermuda/pull/32) - Add handling 3D data for face triangulation and polygon on single plane orthogonal to axis
* [#39](https://github.com/napari/bermuda/pull/39) - Remove consecutive repeated points in polygon
* [conda-forge#29805](https://github.com/conda-forge/staged-recipes/pull/29805) - Add Bermuda to conda-forge

## Retrospection 

When we started talking with the SpatialData team about sponsoring of
work on speedup the Shapes layer, the initial idea was to have a single PR. 
However, after a fast estimation of work required, we decided that it would be 

> much better to work on sequence of PRs. One of the benefits of this

Reviewer: ```suggestion
much better to work on a sequence of PRs. One of the benefits of this
```

using some numpy functions.

However numba is JIT compiler that introduces some delay at first trigger in given session, even when use
cache between runs.

It shows tha try to implement sweeping line algorithm in numba was not successful because of compilation time.

### C++ and Cython 

For gluing C++ code with python we have used Cython. It provides 
Python like code for writing Python API for C++ code.

For example:

```cython
def is_convex(polygon: Sequence[Sequence[float]]) -> bool:
    """ Check if polygon is convex"""
    cdef vector[Point] polygon_vector
    cdef pair[bool, vector[int]] result

    polygon_vector.reserve(len(polygon))
    for point in polygon:
        polygon_vector.push_back(Point(point[0], point[1]))

    return _is_convex(polygon_vector)
```

However for creating numpy array it call numpy API, that introduced performance overhead.
Maybe the better solution could be use [pybind11](https://github.com/pybind/pybind11) 
that offer c++ API for creating numpy array.

The C++ code was much faster than numba one and allow to use Red - Black, that are required for triangulation of shapes. 

### Rust

> During discussion with core-devs we found that we need to increase proficiency with 
> any compiled language to have maintenance capacity. 
> 
> Based on successful stories from different projects we have decided to try Rust.
> When try it on edge triangulation it shows that Rust version is slightly faster than C++ one having memory safety.
> This change may be connected with mentioned above problem with array creation in Cython.
> But even same performance with memory safety is a big win.
> Especially when we meet memory problems in C++ code because of floating point errors. Current error 
> is much more readable than segmentation fault.
> 
> For me, the debugging tools for rust are still slightly worse than C++ one.

Reviewer: ```suggestion
During discussion with core-developers, we found that we had to increase our proficiency with 
any compiled language to have maintenance capacity. 

Based on successful stories from different projects we decided to try Rust.
When trying it on edge triangulation, the Rust version was slightly faster than the C++ implementation and adds memory safety.
This result may be connected with the problem mentioned earlier with array creation in Cython.
However, even same performance with memory safety is a big win.
Also, because when we meet memory problems in C++ code because of floating point errors, we get a cryptic segmentation fault that is hard to trace. The error in Rust is much more readable than this.

For me, the debugging tools for rust are still slightly worse than C++ one.
```
Would have to decide overall whether you want to write this more from your perspective Grzegorz. I think that would be ok.

> ## Future notes

Reviewer: ```suggestion
## Lessons learned
```

> * Start from writing benchmark, instead of custom scripts, it will speedup the process of writing code and testing it.
> * It will be best to have a core-dev funded for reviewing PRs, as early review allow to save time. 
> * Works with multiple PR is muche better than one big PR. It speedup the review process. But require better planning to avoid bottleneck.
> * LLM are better and better in initial review of code. Not all its comments are correct, but it is good starting point. LLM works bad with huge PRs, so it is the next argument to work on smaller PRs.

Reviewer: ```suggestion
* Start with writing a benchmark, instead of custom scripts, it will speedup the process of writing code and testing it.
* It will be best to have a core-dev funded for reviewing PRs, as early review allows to save time. 
* Working with multiple PR is mucher better than one big PR. It speeds up the review process. However, it requires better planning to avoid bottlenecks.
* LLMs are better and better in initial review of code. Not all its comments are correct, but it is a good starting point. LLM works bad with huge PRs, so it is an additional argument to work with smaller PRs.
```
